### ISSUE TITLE
DPC++ front-end compiler allows illegal address space casts.

### ISSUE THREAD

**Issue by bader** [intel/llvm#3172]
I noticed that DPC++ doesn't detect this illegal cast in the FE, but fails to convert it to SPIR-V file:
``` C++
#include <CL/sycl.hpp>
using namespace sycl;
SYCL_EXTERNAL private_ptr<float> bar(global_ptr<float> f) {
  return private_ptr<float>{f.get()};
}
```

``` bash
InvalidModule: Invalid SPIR-V module: Casts from private/local/global address space are allowed only to generic
   %1 = addrspacecast float addrspace(1)* %0 to float*
```

DPC++ FE compiler should diagnose illegal cast between address space attributes.

**Comment by booknouse** [comment]
I meet a similar problem. The error information is:
InvalidModule: Invalid SPIR-V module: Casts from private/local/global address space are allowed only to generic。

I guess that c++ template cause the compile problem.

**Comment by booknouse** [comment]
I find a method to bypass this problem. Move the implemention of template class private_ptr/global_ptr to .h, discard .cpp. Complie will succeed. @bader 

> I noticed that DPC++ doesn't detect this illegal cast in the FE, but fails to convert it to SPIR-V file:
> 
> ```c++
> #include <CL/sycl.hpp>
> using namespace sycl;
> SYCL_EXTERNAL private_ptr<float> bar(global_ptr<float> f) {
>   return private_ptr<float>{f.get()};
> }
> ```
> 
> ```shell
> InvalidModule: Invalid SPIR-V module: Casts from private/local/global address space are allowed only to generic
>    %1 = addrspacecast float addrspace(1)* %0 to float*
> ```
> 
> DPC++ FE compiler should diagnose illegal cast between address space attributes.

**Comment by abagusetty** [comment]
@booknouse @bader I too came across this issue in a large code base and was wondering if there is a way to provide diagnostic info like line number, file, etc either at LLVM level or SPIR-V translator level to identity and address these illegal casts possibly.

**Comment by CaoZhongZ** [comment]
I think both SYCL compiler and SPIR-V should issue a warning instead of error about the issue and proceed with the transform. Basically, if user want to shot in the foot, tools need to give a most efficient way to do it. I came across this issue by assign local accessor to a local_ptr argument, it should be okay and I don't know where compiler lost attribute information.

**Comment by Fznamznon** [comment]
I don't think this particular example reveals any bugs in the FE. This example calls the following constructor of `multi_ptr` class:
https://github.com/intel/llvm/blob/52fad5a98dd802907295205251a55c5b212c03cb/sycl/include/CL/sycl/multi_ptr.hpp#L64
(Please note that SYCL2020 `multi_ptr` class doesn't have such constructor!)
It accepts a pointer in default address space and then **explicitly** converts it to a pointer with named address space. At the point where the constructor is called, implicit cast from pointer with named address space to a pointer with default address space happens. So, in IR we get two casts - from global to generic, which then are optimized to a cast from global to private, that llvm-spirv doesn't allow.
All these operations are allowed by the doc - https://github.com/intel/llvm/blob/sycl/clang/docs/SYCLSupport.rst#id2 .
The bug I've noticed is that FE allows **explicit** casts between named address spaces:
```
__attribute__((opencl_global)) int *GLOB;
__attribute__((opencl_private)) int *NoAS;
GLOB = (__attribute__((opencl_global)) int *)NoAS; // error should be emitted here, now it is not.
``` 
I'm going to fix that.
As a solution to the problem with posted example I suggest to align implementation of `multi_ptr` class with SYCL 2020 spec since it clarifies a lot of things regarding address space handling.

**Comment by Fznamznon** [comment]
https://github.com/intel/llvm/pull/5500 added the warning on attempt to cast from named address space to default. Use -Wsycl-strict to enable it:
```
clang++ -fsycl -Wsycl-strict test.cpp
In file included from test.cpp:1:
In file included from ../include/sycl/CL/sycl.hpp:11:
In file included from ../include/sycl/CL/sycl/accessor.hpp:13:
In file included from ../include/sycl/CL/sycl/buffer.hpp:11:
In file included from ../include/sycl/CL/sycl/detail/buffer_impl.hpp:20:
In file included from ../include/sycl/CL/sycl/types.hpp:56:
../include/sycl/CL/sycl/multi_ptr.hpp:64:47: warning: explicit cast from 'float *' to 'sycl::multi_ptr<float, sycl::access::address_space::private_space>::pointer_t' (aka '__private float *') potentially leads to an invalid address space cast in the resulting code [-Wsycl-strict]
  multi_ptr(ElementType *pointer) : m_Pointer((pointer_t)(pointer)) {
                                              ^~~~~~~~~~~~~~~~~~~~
test.cpp:4:10: note: called by 'bar'
  return private_ptr<float>{f.get()};
         ^
1 warning generated.

```
@booknouse , @abagusetty , @CaoZhongZ Please let me know if it helps.

**Comment by abagusetty** [comment]
@Fznamznon Works for me with the additional flag `-Wsycl-strict`. This can be closed ?
